pkg/controller/deployment_inplace.go

```diff
@@ -0,0 +1,559 @@
+				delete(nodeLabels, v1alpha1.LabelKeyNodeCandidateForUpdate)
+				delete(nodeLabels, v1alpha1.LabelKeyNodeSelectedForUpdate)
+				delete(nodeLabels, v1alpha1.LabelKeyNodeUpdateResult)
+				delete(node.Annotations, v1alpha1.AnnotationKeyMachineUpdateFailedReason)
+				node.ObjectMeta.Labels = nodeLabels
+				node.Spec.Unschedulable = false
+				_, err = dc.targetCoreClient.CoreV1().Nodes().Update(ctx, node, metav1.UpdateOptions{})
+				if err != nil {
+					return fmt.Errorf("failed to uncordon the node %s: %w", node.Name, err)
+				}
+			}
+		}
+	}
+
+	for _, is := range oldISs {
+		// scale down the old machine set to the number of machines which is having the labelselector of the machine set.
+		machines, err := dc.machineLister.List(labels.SelectorFromSet(is.Spec.Selector.MatchLabels))
+		if err != nil {
+			return fmt.Errorf("failed to list machines for machine set %s: %w", is.Name, err)
+		}
+
+		if len(machines) < int(is.Spec.Replicas) {
+			_, _, err := dc.scaleMachineSetAndRecordEvent(ctx, is, int32(len(machines)), deployment) // #nosec G115 (CWE-190) -- value already validated
+			if err != nil {
+				return fmt.Errorf("failed to scale down machine set %s: %w", is.Name, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (dc *controller) reconcileNewMachineSetInPlace(ctx context.Context, oldISs []*v1alpha1.MachineSet, newIS *v1alpha1.MachineSet, deployment *v1alpha1.MachineDeployment) (bool, error) {
+	if (newIS.Spec.Replicas) == (deployment.Spec.Replicas) {
+		// Scaling not required.
+		return false, nil
+	}
+
+	if (newIS.Spec.Replicas) > (deployment.Spec.Replicas) {
+		// Scale down.
+		scaled, _, err := dc.scaleMachineSetAndRecordEvent(ctx, newIS, (deployment.Spec.Replicas), deployment)
+		return scaled, err
+	}
+
+	klog.V(3).Infof("reconcile new machine set %s", newIS.Name)
+
+	addedNewReplicasCount := int32(0)
+
+	for _, is := range oldISs {
+		transferredMachineCount := int32(0)
+		// get the machines for the machine set
+		machines, err := dc.machineLister.List(labels.SelectorFromSet(is.Spec.Selector.MatchLabels))
+		if err != nil {
+			return false, err
+		}
+
+		klog.V(3).Infof("machine in old machine set %s: %d", is.Name, len(machines))
+
+		for _, machine := range machines {
+			nodeName, ok := machine.Labels[v1alpha1.NodeLabelKey]
+			if !ok {
+				return addedNewReplicasCount > 0, fmt.Errorf("node label not found for machine %s", machine.Name)
+			}
+
+			node, err := dc.nodeLister.Get(nodeName)
+			if err != nil {
+				if apierrors.IsNotFound(err) {
+					klog.Warningf("Node %s not found for machine %s", nodeName, machine.Name)
+					continue
+				}
+				return addedNewReplicasCount > 0, fmt.Errorf("failed to get node %s for machine %s: %w", nodeName, machine.Name, err)
+			}
+
+			cond := getMachineCondition(machine, v1alpha1.NodeInPlaceUpdate)
+			if cond == nil || cond.Reason != v1alpha1.UpdateSuccessful || node.Labels[v1alpha1.LabelKeyNodeUpdateResult] != v1alpha1.LabelValueNodeUpdateSuccessful {
```

Comment:
Consider creating a func for this which accepts a condition and a label, and returns a bool